internal/pkg/template/diff/iter.go

```diff
@@ -0,0 +1,108 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package diff
+
+import "gopkg.in/yaml.v3"
+
+type action int
+
+const (
+	match        action = iota
+	modification        = iota
+	deletion            = iota
+	insertion           = iota
+	done                = iota
+)
+
+type tracker[T any] struct {
+	index int
+	data  []T
+}
+
+func newInspector(fromSeq, toSeq []yaml.Node, lcsIndices []lcsIndex) inspector {
+	return inspector{
+		from:       tracker[yaml.Node]{data: fromSeq},
+		to:         tracker[yaml.Node]{data: toSeq},
+		lcsIndices: tracker[lcsIndex]{data: lcsIndices},
+	}
+}
+
+type inspector struct {
+	from       tracker[yaml.Node]
+	to         tracker[yaml.Node]
+	lcsIndices tracker[lcsIndex]
+	currAction action
+}
+
+func (i *inspector) inspect() action {
+	var action action
+	var (
+		commonDone = i.lcsIndices.index >= len(i.lcsIndices.data)
+		fromDone   = i.from.index >= len(i.from.data)
+		toDone     = i.to.index >= len(i.to.data)
+	)
+	if commonDone {
+		switch {
+		case fromDone && toDone:
+			action = done
+		case toDone:
+			// Ex: "a,d,e" -> "a". When the inspector moves to "d" in "from", both common and "to" are done, and "d,e" are considered deleted.
+			action = deletion
+		case fromDone:
+			// Ex: "a" -> "a,d,e". When the inspector moves to "d" in "to", both common and "from" are done, and "d,e" are considered to be inserted.
+			action = insertion
+		default:
+			// Ex: "a,b" -> "a,c". When the inspector moves to index 1 of both list, common is done, and b is modified into c.
```

Comment:
```suggestion
			// Ex: "a,b" -> "a,c". When the inspector moves by 1 index in both lists, common is done, and b is modified into c.
```
is this more accurate?

Reply:
changed it to `// Ex: "a,b" -> "a,c". When the lcsStateMachine moves to (b,c), lcs is done, and b is modified into c.`